packages/edit-site/src/hooks/use-theme-style-variations/use-theme-style-variations-by-property.js

```diff
@@ -0,0 +1,88 @@
+/**
+ * WordPress dependencies
+ */
+import { useMemo } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { mergeBaseAndUserConfigs } from '../../components/global-styles/global-styles-provider';
+import cloneDeep from '../../utils/clone-deep';
+
+/**
+ * Returns a new object with only the properties specified in `properties`.
+ *
+ * @param {Object} object   The object to filter
+ * @param {Object} property The property to filter by
+ * @return {Object} The merged object.
+ */
+export const filterObjectByProperty = ( object, property ) => {
+	if ( ! object ) {
+		return {};
+	}
+
+	const newObject = {};
+	Object.keys( object ).forEach( ( key ) => {
+		if ( key === property ) {
+			newObject[ key ] = object[ key ];
+		} else if ( typeof object[ key ] === 'object' ) {
+			const newFilter = filterObjectByProperty( object[ key ], property );
+			if ( Object.keys( newFilter ).length ) {
+				newObject[ key ] = newFilter;
+			}
+		}
+	} );
+	return newObject;
+};
+
+/**
+ * Returns a new object with only the properties specified in `properties`.
+ *
+ * @param {Object}   props                 Object of hook args.
+ * @param {Object[]} props.styleVariations The theme style variations to filter.
+ * @param {string}   props.styleProperty   The property to filter by.
+ * @param {Function} props.filter          Optional. The filter function to apply to the variations.
+ * @param {Object}   props.baseVariation   Optional. Base or user settings to be updated with variation properties.
+ * @return {Object[]} The merged object.
+ */
+export default function useThemeStyleVariationsByProperty( {
+	styleVariations,
+	styleProperty,
+	filter,
+	baseVariation,
+} ) {
+	return useMemo( () => {
```

Comment:
I measured the performance of this function and `useMemo` increases performance, assuming the same dependencies.

E.g., The initial operation takes `.9ms`, subsequent operations with the same dependencies take `.01ms` or thereabouts